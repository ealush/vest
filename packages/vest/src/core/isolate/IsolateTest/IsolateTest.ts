import { CB, Maybe } from 'vest-utils';
import { TIsolate, Isolate, IsolateKey } from 'vestjs-runtime';

import { TestStatus } from 'IsolateTestStateMachine';
import { TestSeverity } from 'Severity';
import { TFieldName, TGroupName } from 'SuiteResultTypes';
import { AsyncTest, TestFn } from 'TestTypes';
import { VestIsolateType } from 'VestIsolateType';

export type TIsolateTest<
  F extends TFieldName = TFieldName,
  G extends TGroupName = TGroupName
> = TIsolate<CommonTestFields<F, G> & IsolateTestPayload>;

export function IsolateTest<
  F extends TFieldName = TFieldName,
  G extends TGroupName = TGroupName
>(
  callback: CB,
  input: CommonTestFields<F, G>,
  key?: IsolateKey
): TIsolateTest<F, G> {
  const payload: IsolateTestPayload = {
    ...IsolateTestBase(),
    fieldName: input.fieldName,
    testFn: input.testFn,
  };

  if (input.groupName) {
    payload.groupName = input.groupName;
  }

  if (input.message) {
    payload.message = input.message;
  }
  const isolate = Isolate.create<IsolateTestPayload>(
    VestIsolateType.Test,
    callback,
    payload,
    key ?? null
  );

  return isolate as TIsolateTest<F, G>;
}

export function IsolateTestBase() {
  return {
    severity: TestSeverity.Error,
    status: TestStatus.UNTESTED,
    abortController: new AbortController(),
  };
}

export type IsolateTestPayload<
  F extends TFieldName = TFieldName,
  G extends TGroupName = TGroupName
> = CommonTestFields<F, G> & {
  severity: TestSeverity;
  status: TestStatus;
  asyncTest?: AsyncTest;
  abortController: AbortController;
};

type CommonTestFields<
  F extends TFieldName = TFieldName,
  G extends TGroupName = TGroupName
> = {
  message?: Maybe<string>;
  groupName?: G;
  fieldName: F;
  testFn: TestFn;
};
