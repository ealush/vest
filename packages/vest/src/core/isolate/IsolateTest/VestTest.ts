import { invariant, isPromise, optionalFunctionValue } from 'vest-utils';

import type { TIsolateTest } from 'IsolateTest';
import {
  TestAction,
  TestStateMachineAction,
  TestStatus,
  createTestStateMachine,
} from 'IsolateTestStateMachine';
import { TestSeverity } from 'Severity';
import { TFieldName, TGroupName } from 'SuiteResultTypes';

const TestStateMachine = createTestStateMachine();

export class VestTest {
  // Read

  static getData<
    F extends TFieldName = TFieldName,
    G extends TGroupName = TGroupName
  >(test: TIsolateTest<F, G>) {
    invariant(test.data);
    return test.data;
  }

  static warns(test: TIsolateTest): boolean {
    return VestTest.getData(test).severity === TestSeverity.Warning;
  }

  static isPending(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.PENDING);
  }

  static isOmitted(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.OMITTED);
  }

  static isUntested(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.UNTESTED);
  }

  static isFailing(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.FAILED);
  }

  static isCanceled(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.CANCELED);
  }

  static isSkipped(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.SKIPPED);
  }

  static isPassing(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.PASSING);
  }

  static isWarning(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.WARNING);
  }

  static hasFailures(test: TIsolateTest): boolean {
    return VestTest.isFailing(test) || VestTest.isWarning(test);
  }

  static isNonActionable(test: TIsolateTest): boolean {
    return (
      VestTest.isSkipped(test) ||
      VestTest.isOmitted(test) ||
      VestTest.isCanceled(test)
    );
  }

  static isTested(test: TIsolateTest): boolean {
    return VestTest.hasFailures(test) || VestTest.isPassing(test);
  }

  static awaitsResolution(test: TIsolateTest): boolean {
    // Is the test in a state where it can still be run, or complete running
    // and its final status is indeterminate?
    return (
      VestTest.isSkipped(test) ||
      VestTest.isUntested(test) ||
      VestTest.isPending(test)
    );
  }

  static isAsyncTest(test: TIsolateTest): boolean {
    return isPromise(VestTest.getData(test).asyncTest);
  }

  static statusEquals(test: TIsolateTest, status: TestStatus): boolean {
    return VestTest.getData(test).status === status;
  }

  // Mutate

  static setPending(test: TIsolateTest) {
    VestTest.setStatus(test, TestStatus.PENDING);
  }

  static fail(test: TIsolateTest): void {
    VestTest.setStatus(
      test,
      VestTest.warns(test) ? TestStatus.WARNING : TestStatus.FAILED
    );
  }

  static pass(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.PASSING);
  }

  static warn(test: TIsolateTest): void {
    VestTest.setData(test, current => ({
      ...current,
      severity: TestSeverity.Warning,
    }));
  }

  static setData(
    test: TIsolateTest,
    setter:
      | ((current: TIsolateTest['data']) => TIsolateTest['data'])
      | TIsolateTest['data']
  ): void {
    test.data = optionalFunctionValue(setter, VestTest.getData(test));
  }

  static skip(test: TIsolateTest, force?: boolean): void {
    // Without this force flag, the test will be marked as skipped even if it is pending.
    // This means that it will not be counted in "allIncomplete" and its done callbacks
    // will not be called, or will be called prematurely.
    // What this mostly say is that when we have a pending test for one field, and we then
    // start typing in a different field - the pending test will be canceled, which
    // is usually an unwanted behavior.
    // The only scenario in which we DO want to cancel the async test regardless
    // is when we specifically skip a test with `skipWhen`, which is handled by the
    // "force" boolean flag.
    // I am not a fan of this flag, but it gets the job done.
    VestTest.setStatus(test, TestStatus.SKIPPED, force);
  }

  static cancel(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.CANCELED);
    VestTest.getData(test).abortController.abort(TestStatus.CANCELED);
  }

  static omit(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.OMITTED);
  }

  static reset(test: TIsolateTest): void {
    VestTest.setStatus(test, TestAction.RESET);
  }

  static setStatus(
    test: TIsolateTest,
    status: TestStateMachineAction,
    payload?: any
  ): void {
    VestTest.setData(test, current => ({
      ...current,
      status: TestStateMachine.staticTransition(
        current.status,
        status,
        payload
      ),
    }));
  }
}
