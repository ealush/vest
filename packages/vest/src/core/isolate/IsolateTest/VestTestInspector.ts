import { isPromise } from 'vest-utils';

import type { TIsolateTest } from 'IsolateTest';
import { TestStatus } from 'IsolateTestStateMachine';
import { TestSeverity } from 'Severity';

export class VestTestInspector {
  static warns(test: TIsolateTest): boolean {
    return test.severity === TestSeverity.Warning;
  }

  static isPending(test: TIsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.PENDING);
  }

  static isOmitted(test: TIsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.OMITTED);
  }

  static isUntested(test: TIsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.UNTESTED);
  }

  static isFailing(test: TIsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.FAILED);
  }

  static isCanceled(test: TIsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.CANCELED);
  }

  static isSkipped(test: TIsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.SKIPPED);
  }

  static isPassing(test: TIsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.PASSING);
  }

  static isWarning(test: TIsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.WARNING);
  }

  static hasFailures(test: TIsolateTest): boolean {
    return (
      VestTestInspector.isFailing(test) || VestTestInspector.isWarning(test)
    );
  }

  static isNonActionable(test: TIsolateTest): boolean {
    return (
      VestTestInspector.isSkipped(test) ||
      VestTestInspector.isOmitted(test) ||
      VestTestInspector.isCanceled(test)
    );
  }

  static isTested(test: TIsolateTest): boolean {
    return (
      VestTestInspector.hasFailures(test) || VestTestInspector.isPassing(test)
    );
  }

  static awaitsResolution(test: TIsolateTest): boolean {
    // Is the test in a state where it can still be run, or complete running
    // and its final status is indeterminate?
    return (
      VestTestInspector.isSkipped(test) ||
      VestTestInspector.isUntested(test) ||
      VestTestInspector.isPending(test)
    );
  }

  static isAsyncTest(test: TIsolateTest): boolean {
    return isPromise(test.asyncTest);
  }

  static statusEquals(test: TIsolateTest, status: TestStatus): boolean {
    return test.status === status;
  }
}
