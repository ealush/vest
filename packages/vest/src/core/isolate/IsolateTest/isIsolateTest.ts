import { Maybe, invariant } from 'vest-utils';
import { TIsolate, IsolateSelectors } from 'vestjs-runtime';

import type { TIsolateTest } from 'IsolateTest';
import { TFieldName, TGroupName } from 'SuiteResultTypes';
import { VestIsolateType } from 'VestIsolateType';

export function isIsolateTest(
  isolate?: Maybe<TIsolate>
): isolate is TIsolateTest {
  return IsolateSelectors.isIsolateType<TIsolateTest>(
    isolate,
    VestIsolateType.Test
  );
}

export function isIsolateTestX(
  isolate: TIsolate
): asserts isolate is TIsolateTest {
  invariant(isIsolateTest(isolate));
}

export function castIsolateTest<
  F extends TFieldName = TFieldName,
  G extends TGroupName = TGroupName
>(isolate: TIsolate): TIsolateTest<F, G> {
  isIsolateTestX(isolate);

  return isolate as TIsolateTest<F, G>;
}
