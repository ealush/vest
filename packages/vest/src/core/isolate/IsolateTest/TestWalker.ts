import { Walker, Isolate, VestRuntime } from 'vest-runtime';

import { IsolateTest } from 'IsolateTest';
import { TFieldName, TGroupName } from 'SuiteResultTypes';
import matchingFieldName from 'matchingFieldName';

type MaybeRoot = Isolate | null;

export class TestWalker {
  static defaultRoot() {
    return VestRuntime.useAvailableRoot();
  }

  static hasNoTests(root: MaybeRoot = TestWalker.defaultRoot()): boolean {
    if (!root) return true;
    return !Walker.has(root, IsolateTest.is);
  }

  static someIncompleteTests(
    predicate: (test: IsolateTest) => boolean,
    root: MaybeRoot = TestWalker.defaultRoot()
  ): boolean {
    if (!root) return false;
    return Walker.some(
      root,
      isolate => {
        IsolateTest.isX(isolate);

        return isolate.isPending() && predicate(isolate);
      },
      IsolateTest.is
    );
  }

  static someTests(
    predicate: (test: IsolateTest) => boolean,
    root: MaybeRoot = TestWalker.defaultRoot()
  ): boolean {
    if (!root) return false;
    return Walker.some(
      root,
      isolate => {
        IsolateTest.isX(isolate);

        return predicate(isolate);
      },
      IsolateTest.is
    );
  }

  static everyTest(
    predicate: (test: IsolateTest) => boolean,
    root: MaybeRoot = TestWalker.defaultRoot()
  ): boolean {
    if (!root) return false;
    return Walker.every(
      root,
      isolate => {
        IsolateTest.isX(isolate);

        return predicate(isolate);
      },
      IsolateTest.is
    );
  }

  static walkTests<F extends TFieldName, G extends TGroupName>(
    callback: (test: IsolateTest<F, G>, breakout: () => void) => void,
    root: MaybeRoot = TestWalker.defaultRoot()
  ): void {
    if (!root) return;
    Walker.walk(
      root,
      (isolate, breakout) => {
        callback(IsolateTest.cast<F, G>(isolate), breakout);
      },
      IsolateTest.is
    );
  }

  static hasRemainingTests(fieldName?: TFieldName): boolean {
    return TestWalker.someIncompleteTests(testObject => {
      if (fieldName) {
        return matchingFieldName(testObject, fieldName);
      }
      return true;
    });
  }

  static pluckTests(
    predicate: (test: IsolateTest) => boolean,
    root: MaybeRoot = TestWalker.defaultRoot()
  ): void {
    if (!root) return;
    Walker.pluck(
      root,
      isolate => {
        IsolateTest.isX(isolate);

        return predicate(isolate);
      },
      IsolateTest.is
    );
  }

  static resetField(fieldName: TFieldName): void {
    TestWalker.walkTests(testObject => {
      if (matchingFieldName(testObject, fieldName)) {
        testObject.reset();
      }
    }, TestWalker.defaultRoot());
  }

  static removeTestByFieldName(
    fieldName: TFieldName,
    root: MaybeRoot = TestWalker.defaultRoot()
  ): void {
    TestWalker.pluckTests(testObject => {
      return matchingFieldName(testObject, fieldName);
    }, root);
  }
}
