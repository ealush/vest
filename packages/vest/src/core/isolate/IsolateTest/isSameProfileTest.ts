import { TIsolateTest } from 'IsolateTest';
import matchingFieldName from 'matchingFieldName';

export function isSameProfileTest(
  testObject1: TIsolateTest,
  testObject2: TIsolateTest
): boolean {
  return (
    matchingFieldName(testObject1, testObject2.fieldName) &&
    testObject1.groupName === testObject2.groupName &&
    testObject1.key === testObject2.key
  );
}
