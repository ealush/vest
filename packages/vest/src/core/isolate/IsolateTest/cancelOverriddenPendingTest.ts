import { TIsolateTest } from 'IsolateTest';
import { VestTest, VestTest } from 'VestTest';
import { isSameProfileTest } from 'isSameProfileTest';

export default function cancelOverriddenPendingTest(
  prevRunTestObject: TIsolateTest,
  currentRunTestObject: TIsolateTest
): void {
  if (
    currentRunTestObject !== prevRunTestObject &&
    isSameProfileTest(prevRunTestObject, currentRunTestObject) &&
    VestTest.isPending(prevRunTestObject)
  ) {
    VestTest.cancel(prevRunTestObject);
  }
}
