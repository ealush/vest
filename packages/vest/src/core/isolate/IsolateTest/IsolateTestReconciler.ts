import { ErrorStrings } from 'ErrorStrings';
import { Reconciler } from 'vest-runtime';
import type { Isolate } from 'vest-runtime';
import { deferThrow, isNullish, text } from 'vest-utils';

import { IsolateTest } from 'IsolateTest';
import cancelOverriddenPendingTest from 'cancelOverriddenPendingTest';
import { isSameProfileTest } from 'isSameProfileTest';
import { useVerifyTestRun } from 'verifyTestRun';

// @vx-allow use-use
export function IsolateTestReconciler(
  currentNode: Isolate,
  historyNode: Isolate | null
): Isolate {
  // Start by verifying params
  if (!IsolateTest.is(currentNode)) {
    // This is unreachable, since this function should only be called with IsolateTest nodes
    return currentNode;
  }

  if (isNullish(historyNode)) {
    return handleNoHistoryNode(currentNode);
  }

  if (!IsolateTest.is(historyNode)) {
    return currentNode;
  }

  const reconcilerOutput = usePickNode(historyNode, currentNode);

  cancelOverriddenPendingTestOnTestReRun(
    reconcilerOutput,
    currentNode,
    historyNode
  );

  return reconcilerOutput;
}

// eslint-disable-next-line max-statements

function nodeReorderDetected(
  newNode: IsolateTest,
  prevNode?: Isolate
): boolean {
  return !!IsolateTest.is(prevNode) && !isSameProfileTest(prevNode, newNode);
}

function handleCollision(
  newNode: IsolateTest,
  prevNode?: Isolate
): IsolateTest {
  if (newNode.usesKey()) {
    return IsolateTest.cast(Reconciler.handleIsolateNodeWithKey(newNode));
  }

  if (nodeReorderDetected(newNode, prevNode)) {
    return onNodeReorder(newNode, prevNode);
  }

  if (!IsolateTest.is(prevNode)) {
    // I believe we cannot actually reach this point.
    // Because it should already be handled by nodeReorderDetected.
    return newNode;
  }

  // FIXME: May-13-2023
  // This may not be the most ideal solution.
  // In short: if the node was omitted in the previous run,
  // we want to re-evaluate it. The reason is that we may incorrectly
  // identify it is "optional" because it was omitted in the previous run.
  // There may be a better way to handle this. Need to revisit this.
  if (prevNode.isOmitted()) {
    return newNode;
  }

  return prevNode;
}

function onNodeReorder(newNode: IsolateTest, prevNode?: Isolate): IsolateTest {
  throwTestOrderError(newNode, prevNode);
  Reconciler.removeAllNextNodesInIsolate();
  return newNode;
}

function usePickNode(
  historyNode: IsolateTest,
  currentNode: IsolateTest
): Isolate {
  const collisionResult = handleCollision(currentNode, historyNode);

  return useVerifyTestRun(currentNode, collisionResult);
}

function handleNoHistoryNode(testNode: IsolateTest): IsolateTest {
  if (testNode.usesKey()) {
    return IsolateTest.cast(Reconciler.handleIsolateNodeWithKey(testNode));
  }

  return testNode;
}

function cancelOverriddenPendingTestOnTestReRun(
  nextNode: Isolate,
  currentNode: Isolate,
  prevTestObject: IsolateTest
) {
  if (nextNode === currentNode && IsolateTest.is(currentNode)) {
    cancelOverriddenPendingTest(prevTestObject, currentNode);
  }
}

function throwTestOrderError(
  newNode: IsolateTest,
  prevNode: Isolate | undefined
): void {
  if (newNode.shouldAllowReorder()) {
    return;
  }

  deferThrow(
    text(ErrorStrings.TESTS_CALLED_IN_DIFFERENT_ORDER, {
      fieldName: newNode.fieldName,
      prevName: IsolateTest.is(prevNode) ? prevNode.fieldName : undefined,
    })
  );
}
