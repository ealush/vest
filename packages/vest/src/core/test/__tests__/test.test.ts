import { faker } from '@faker-js/faker';
import { text } from 'vest-utils';
import { IsolateInspector } from 'vestjs-runtime';

import { TestPromise } from '../../../testUtils/testPromise';

import { ErrorStrings } from 'ErrorStrings';
import { TIsolateTest } from 'IsolateTest';
import { VestTestInspector } from 'VestTestInspector';
import { enforce } from 'vest';
import * as vest from 'vest';

let testObject: TIsolateTest;

describe("Test Vest's `test` function", () => {
  describe('test callbacks', () => {
    describe('Warn hook', () => {
      it('Should be marked as warning when the warn hook gets called', () => {
        vest.create(() => {
          testObject = vest.test(
            faker.random.word(),
            faker.lorem.sentence(),
            () => {
              vest.warn();
            }
          );
        })();
        expect(VestTestInspector.warns(testObject)).toBe(true);
      });
    });

    describe('Sync', () => {
      it('Should be marked as failed after a thrown error', () => {
        vest.create(() => {
          testObject = vest.test(
            faker.random.word(),
            faker.lorem.sentence(),
            () => {
              throw new Error();
            }
          );
        })();
        expect(VestTestInspector.isFailing(testObject)).toBe(true);
      });

      it('Should be marked as failed for an explicit false return', () => {
        vest.create(() => {
          vest.test(faker.random.word(), faker.lorem.sentence(), () => false);
        })();
        expect(VestTestInspector.isFailing(testObject)).toBe(true);
      });

      describe('Thrown with a message', () => {
        describe('When field has a message', () => {
          it("Should use field's own message", () => {
            const res = vest.create(() => {
              vest.test('field_with_message', 'some_field_message', () => {
                failWithString();
              });
              vest.test(
                'warning_field_with_message',
                'some_field_message',
                () => {
                  vest.warn();
                  failWithString();
                }
              );
            })();

            expect(res.getErrors('field_with_message')).toEqual([
              'some_field_message',
            ]);
            expect(res.tests['field_with_message'].errors).toEqual([
              'some_field_message',
            ]);
            expect(res.getWarnings('warning_field_with_message')).toEqual([
              'some_field_message',
            ]);
            expect(res.tests['warning_field_with_message'].warnings).toEqual([
              'some_field_message',
            ]);
          });
        });
        describe('When field does not have a message', () => {
          it('Should use message from enforce().message()', () => {
            const res = vest.create(() => {
              vest.test('field_without_message', () => {
                enforce(100).message('some_field_message').equals(0);
              });
            })();

            expect(res.getErrors('field_without_message')).toEqual([
              'some_field_message',
            ]);
          });
          it('Should use message from thrown error', () => {
            const res = vest.create(() => {
              vest.test('field_without_message', () => {
                failWithString();
              });
              vest.test('warning_field_without_message', () => {
                vest.warn();
                failWithString();
              });
            })();

            expect(res.getErrors('field_without_message')).toEqual([
              'I fail with a message',
            ]);
            expect(res.tests['field_without_message'].errors).toEqual([
              'I fail with a message',
            ]);
            expect(res.getWarnings('warning_field_without_message')).toEqual([
              'I fail with a message',
            ]);
            expect(res.tests['warning_field_without_message'].warnings).toEqual(
              ['I fail with a message']
            );
          });
        });
      });
    });

    describe('async', () => {
      it('Should be marked as failed when a returned promise rejects', () =>
        TestPromise(done => {
          vest.create(() => {
            testObject = vest.test(
              faker.random.word(),
              faker.lorem.sentence(),
              () =>
                new Promise((_, reject) => {
                  expect(VestTestInspector.isFailing(testObject)).toBe(false);
                  setTimeout(reject, 300);
                })
            );
            expect(VestTestInspector.isFailing(testObject)).toBe(false);
            setTimeout(() => {
              expect(VestTestInspector.isFailing(testObject)).toBe(true);
              done();
            }, 310);
          })();
        }));
    });
  });

  describe('test params', () => {
    let testObject: TIsolateTest;
    it('creates a test without a message and without a key', () => {
      vest.create(() => {
        testObject = vest.test('field_name', () => undefined);
      })();
      expect(testObject.fieldName).toBe('field_name');
      expect(testObject.key).toBeNull();
      expect(testObject.message).toBeUndefined();
      expect(testObject).toMatchSnapshot();
    });

    it('creates a test without a key', () => {
      vest.create(() => {
        testObject = vest.test(
          'field_name',
          'failure message',
          () => undefined
        );
      })();
      expect(testObject.fieldName).toBe('field_name');
      expect(testObject.key).toBeNull();
      expect(testObject.message).toBe('failure message');
      expect(testObject).toMatchSnapshot();
    });

    it('creates a test without a message and with a key', () => {
      vest.create(() => {
        testObject = vest.test('field_name', () => undefined, 'keyboardcat');
      })();
      expect(testObject.fieldName).toBe('field_name');
      expect(testObject.key).toBe('keyboardcat');
      expect(testObject.message).toBeUndefined();
      expect(testObject).toMatchSnapshot();
    });

    it('creates a test with a message and with a key', () => {
      vest.create(() => {
        testObject = vest.test(
          'field_name',
          'failure message',
          () => undefined,
          'keyboardcat'
        );
      })();
      expect(testObject.fieldName).toBe('field_name');
      expect(testObject.key).toBe('keyboardcat');
      expect(testObject.message).toBe('failure message');
      expect(IsolateInspector.dump(testObject)).toMatchInlineSnapshot(
        `"{"children":[],"keys":{},"type":"Test","severity":"error","status":"PASSING","fieldName":"field_name","message":"failure message","key":"keyboardcat"}"`
      );
    });

    it('throws when field name is not a string', () => {
      const control = jest.fn();
      vest.create(() => {
        // @ts-ignore
        expect(() => vest.test(undefined, () => undefined)).toThrow(
          text(ErrorStrings.INVALID_PARAM_PASSED_TO_FUNCTION, {
            fn_name: 'test',
            param: 'fieldName',
            expected: 'string',
          })
        );
        // @ts-expect-error
        expect(() => vest.test(null, 'error message', () => undefined)).toThrow(
          text(ErrorStrings.INVALID_PARAM_PASSED_TO_FUNCTION, {
            fn_name: 'test',
            param: 'fieldName',
            expected: 'string',
          })
        );
        expect(() =>
          // @ts-expect-error
          vest.test(null, 'error message', () => undefined, 'key')
        ).toThrow(
          text(ErrorStrings.INVALID_PARAM_PASSED_TO_FUNCTION, {
            fn_name: 'test',
            param: 'fieldName',
            expected: 'string',
          })
        );
        control();
      })();
      expect(control).toHaveBeenCalled();
    });

    it('throws when callback is not a function', () => {
      const control = jest.fn();
      vest.create(() => {
        // @ts-expect-error
        expect(() => vest.test('x')).toThrow(
          text(ErrorStrings.INVALID_PARAM_PASSED_TO_FUNCTION, {
            fn_name: 'test',
            param: 'callback',
            expected: 'function',
          })
        );
        // @ts-expect-error
        expect(() => vest.test('x', 'msg', undefined)).toThrow(
          text(ErrorStrings.INVALID_PARAM_PASSED_TO_FUNCTION, {
            fn_name: 'test',
            param: 'callback',
            expected: 'function',
          })
        );
        // @ts-expect-error
        expect(() => vest.test('x', 'msg', undefined, 'key')).toThrow(
          text(ErrorStrings.INVALID_PARAM_PASSED_TO_FUNCTION, {
            fn_name: 'test',
            param: 'callback',
            expected: 'function',
          })
        );
        control();
      })();
      expect(control).toHaveBeenCalled();
    });
  });
});

function failWithString(msg?: string) {
  throw msg ?? 'I fail with a message';
}
