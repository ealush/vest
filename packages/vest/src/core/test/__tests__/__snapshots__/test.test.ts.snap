// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Vest's \`test\` function test params creates a test without a key 1`] = `
{
  "$type": "Test",
  "allowReorder": undefined,
  "children": null,
  "data": {
    "fieldName": "field_name",
    "message": "failure message",
    "severity": "error",
    "status": "PASSING",
    "testFn": [Function],
  },
  "key": null,
  "keys": null,
  "output": undefined,
  "parent": {
    "$type": "Suite",
    "allowReorder": undefined,
    "children": [
      [Circular],
    ],
    "data": {
      "optional": {},
    },
    "key": null,
    "keys": null,
    "output": {
      "done": [Function],
      "errorCount": 0,
      "errors": [],
      "getError": [Function],
      "getErrors": [Function],
      "getErrorsByGroup": [Function],
      "getWarning": [Function],
      "getWarnings": [Function],
      "getWarningsByGroup": [Function],
      "groups": {},
      "hasErrors": [Function],
      "hasErrorsByGroup": [Function],
      "hasWarnings": [Function],
      "hasWarningsByGroup": [Function],
      "isPending": [Function],
      "isValid": [Function],
      "isValidByGroup": [Function],
      "pendingCount": 0,
      "suiteName": undefined,
      "testCount": 1,
      "tests": {
        "field_name": SummaryBase {
          "errorCount": 0,
          "errors": [],
          "pendingCount": 0,
          "testCount": 1,
          "valid": true,
          "warnCount": 0,
          "warnings": [],
        },
      },
      "valid": true,
      "warnCount": 0,
      "warnings": [],
    },
    "parent": null,
  },
}
`;

exports[`Test Vest's \`test\` function test params creates a test without a message and with a key 1`] = `
{
  "$type": "Test",
  "allowReorder": undefined,
  "children": null,
  "data": {
    "fieldName": "field_name",
    "severity": "error",
    "status": "PASSING",
    "testFn": [Function],
  },
  "key": "keyboardcat",
  "keys": null,
  "output": undefined,
  "parent": {
    "$type": "Suite",
    "allowReorder": undefined,
    "children": [
      [Circular],
    ],
    "data": {
      "optional": {},
    },
    "key": null,
    "keys": {
      "keyboardcat": [Circular],
    },
    "output": {
      "done": [Function],
      "errorCount": 0,
      "errors": [],
      "getError": [Function],
      "getErrors": [Function],
      "getErrorsByGroup": [Function],
      "getWarning": [Function],
      "getWarnings": [Function],
      "getWarningsByGroup": [Function],
      "groups": {},
      "hasErrors": [Function],
      "hasErrorsByGroup": [Function],
      "hasWarnings": [Function],
      "hasWarningsByGroup": [Function],
      "isPending": [Function],
      "isValid": [Function],
      "isValidByGroup": [Function],
      "pendingCount": 0,
      "suiteName": undefined,
      "testCount": 1,
      "tests": {
        "field_name": SummaryBase {
          "errorCount": 0,
          "errors": [],
          "pendingCount": 0,
          "testCount": 1,
          "valid": true,
          "warnCount": 0,
          "warnings": [],
        },
      },
      "valid": true,
      "warnCount": 0,
      "warnings": [],
    },
    "parent": null,
  },
}
`;

exports[`Test Vest's \`test\` function test params creates a test without a message and without a key 1`] = `
{
  "$type": "Test",
  "allowReorder": undefined,
  "children": null,
  "data": {
    "fieldName": "field_name",
    "severity": "error",
    "status": "PASSING",
    "testFn": [Function],
  },
  "key": null,
  "keys": null,
  "output": undefined,
  "parent": {
    "$type": "Suite",
    "allowReorder": undefined,
    "children": [
      [Circular],
    ],
    "data": {
      "optional": {},
    },
    "key": null,
    "keys": null,
    "output": {
      "done": [Function],
      "errorCount": 0,
      "errors": [],
      "getError": [Function],
      "getErrors": [Function],
      "getErrorsByGroup": [Function],
      "getWarning": [Function],
      "getWarnings": [Function],
      "getWarningsByGroup": [Function],
      "groups": {},
      "hasErrors": [Function],
      "hasErrorsByGroup": [Function],
      "hasWarnings": [Function],
      "hasWarningsByGroup": [Function],
      "isPending": [Function],
      "isValid": [Function],
      "isValidByGroup": [Function],
      "pendingCount": 0,
      "suiteName": undefined,
      "testCount": 1,
      "tests": {
        "field_name": SummaryBase {
          "errorCount": 0,
          "errors": [],
          "pendingCount": 0,
          "testCount": 1,
          "valid": true,
          "warnCount": 0,
          "warnings": [],
        },
      },
      "valid": true,
      "warnCount": 0,
      "warnings": [],
    },
    "parent": null,
  },
}
`;
